src/main/java/org/mastodon/revised/mamut/feature/DefaultMamutFeatureComputerService.java

```diff
@@ -0,0 +1,255 @@
+package org.mastodon.revised.mamut.feature;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.mastodon.graph.algorithm.TopologicalSort;
+import org.mastodon.graph.object.ObjectEdge;
+import org.mastodon.graph.object.ObjectGraph;
+import org.mastodon.graph.object.ObjectVertex;
+import org.mastodon.revised.mamut.ProgressListener;
+import org.mastodon.revised.model.feature.Feature;
+import org.mastodon.revised.model.feature.FeatureComputer;
+import org.mastodon.revised.model.feature.FeatureComputerService;
+import org.mastodon.revised.model.mamut.Model;
+import org.scijava.InstantiableException;
+import org.scijava.app.StatusService;
+import org.scijava.log.LogService;
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.Plugin;
+import org.scijava.plugin.PluginInfo;
+import org.scijava.plugin.PluginService;
+import org.scijava.service.AbstractService;
+
+@Plugin( type = FeatureComputerService.class )
+public class DefaultMamutFeatureComputerService extends AbstractService implements FeatureComputerService< Model >
+{
+
+	@Parameter
+	private PluginService pluginService;
+
+	@Parameter
+	private LogService logService;
+
+	@Parameter
+	private StatusService status;
+
+	/**
+	 * Feature computers of any type.
+	 */
+	private final Map< String, FeatureComputer< Model > > availableFeatureComputers = new HashMap<>();
+
+	/**
+	 * Available spot feature computers.
+	 */
+	private Set< String > availableSpotFeatureComputers;
+
+	/**
+	 * Available link feature computers.
+	 */
+	private Set< String > availableLinkFeatureComputers;
+
+	public DefaultMamutFeatureComputerService()
+	{}
+
+	@Override
+	public void initialize()
+	{
+		super.initialize();
+		initializeAvailableSpotFeatureComputers();
+		initializeAvailableLinkFeatureComputers();
+	}
+
+	@Override
+	public final Set< String > getAvailableVertexFeatureComputers()
+	{
+		return availableSpotFeatureComputers;
+	}
+
+	@Override
+	public Set< String > getAvailableEdgeFeatureComputers()
+	{
+		return availableLinkFeatureComputers;
+	}
+
+	@Override
+	public boolean compute( final Model model, final Set< String > computerNames, final ProgressListener progressListener )
+	{
+		final ObjectGraph< FeatureComputer< Model > > dependencyGraph = getDependencyGraph( computerNames );
+		final TopologicalSort< ObjectVertex< FeatureComputer< Model > >, ObjectEdge< FeatureComputer< Model > > > sorter
+			= new TopologicalSort<>( dependencyGraph );
+
+		if (sorter.hasFailed())
+		{
+			logService.error( "Could not compute features using  " + computerNames +
+					" as they have a circular dependency." );
+			progressListener.showStatus( "Circular dependency!" );
+			return false;
+		}
+
+		final long start = System.currentTimeMillis();
+
+		model.getGraphFeatureModel().clear();
```

Comment:
This will cause problems with `manual` features? Complete `clear()` seems wrong here?

Reply:
True, noted.